### PR TITLE
Update to only fetch transactions for blocks when necessary.

### PR DESCRIPTION
**Pull request by akhila-raju**
Implements #49.

### CODE REVIEW

**Comment by raulk** [comment]
@akhila-raju a few comments:
* Where is the new logic being tested? Ideally we'd be having tests that verify that `web3.eth.getBlock()` was called with the second argument set to true/false depending on the cases. I think you can use [`jest.spyOn()`](https://facebook.github.io/jest/docs/en/jest-object#jestspyonobject-methodname) for this.
* package-lock.json is not a file we want in the repo. Perhaps you're using npm/npx instead of yarn?

File: src/model/EthqlQuery.ts [hunk 1]
```diff
@@ -121,6 +130,15 @@ class EthqlQuery {
   public health() {
     return 'ok';
   }
+
+  public queryContainsTxs(info: GraphQLResolveInfo): boolean {
+    const fields = getFieldList(info);
+    return (
+      !!fields.find(
+        a => a.includes('transactions') || a.includes('transactionsInvolving') || a.includes('transactionsRoles'),
```
Comment (raulk) [inline src/model/EthqlQuery.ts]: I suspect that the list of inner fields that triggers transaction retrieval should be a constant of some kind somewhere.

File: src/model/core/EthqlBlock.ts [hunk 2]
```diff
@@ -65,8 +66,10 @@ class EthqlBlock implements EthqlBlock {
     return this._transactions.length;
   }
 
-  public transactionAt(args): EthqlTransaction {
-    return this._transactions[args.index];
+  public async transactionAt(args, { web3 }: EthqlContext): Promise<EthqlTransaction> {
+    if (args.index < 0) { return null; }
```
Comment (raulk) [inline src/model/core/EthqlBlock.ts]: Good catch adding a guard here. Maybe we should add a test for this case?

File: src/__tests__/core/blocksRange.test.ts [hunk 3]
```diff
@@ -243,3 +244,20 @@ test('blocksRange: same block hash provided twice returns one block', async () =
   const result = await execQuery(query);
   expect(result).toEqual(expected);
 });
+
+test('blocksRange: web3.eth.getBlock called with 2nd param false value', async () => {
```
Comment (raulk) [inline src/__tests__/core/blocksRange.test.ts]: This test is not actually spying on `web3.eth.getBlock`, right? From what I see, the test checks that `queryContainsTxs` is invoked and returns false, but still the test is not covering `getBlock(..., true/false)` which is the assertion we want to make.

Comment (raulk) [reply src/__tests__/core/blocksRange.test.ts]: Same in the other variants (blocks, block).

File: src/model/core/EthqlBlock.ts (hunk 2 above)
Comment (akhila-raju) [reply src/model/core/EthqlBlock.ts]: SGTM

File: src/model/EthqlQuery.ts (hunk 1 above)
Comment (akhila-raju) [reply src/model/EthqlQuery.ts]: SGTM - will change

**Comment by raulk** [comment]
@akhila-raju pushed a few commits here to:

* Consider only immediate children to avoid false positives when transactions are requested for nested blocks, not the block at hand (e.g. parent, ommers).
* Remove dependency on `graphql-list-fields`.
* Reorganize related tests into a dedicated suite.
* Field whitelist is now in a static field.

Further refactoring is required in the model, as it's questionable for this logic to be in the `EthqlQuery` root query, it should be in the `EthqlBlock` class so it can be reused across other fields that yield a Block (parent, ommers, and others down the line).

**Pull request merged**